File: Python/DesignPatterns/Builder_Design_Pattern.py
# Product

class Student:
    def __init__(self, builder: "Student.StudentBuilder"):
        self.name = builder.name
        self.roll_number = builder.roll_number
        self.age = builder.age
        self.address = builder.address

    class StudentBuilder:
        def __init__(self, name: str):
            self.name = name  # Mandatory
            self.roll_number: int = 0  # Optional Parameter
            self.age: int = 0  # Optional Parameter
            self.address: str = ""  # Optional Parameter

        def set_roll_number(self, roll_number: int) -> "Student.StudentBuilder":
            self.roll_number = roll_number
            return self

        def set_age(self, age: int) -> "Student.StudentBuilder":
            self.age = age
            return self

        def set_address(self, address: str) -> "Student.StudentBuilder":
            self.address = address
            return self

        def build(self) -> "Student":
            return Student(self)


if __name__ == "__main__":
    student = (
        Student.StudentBuilder("pradeep")
        .set_address("dada")
        .set_age(123)
        .set_address("dads")
        .build()
    )

    # Is modification of student possible..

File: JavaScript/DesignPatterns/Builder_Design_Pattern.js
// Product

class StudentBuilder {
    constructor(name) {
        this.name = name; // Mandatory
        this.rollNumber = undefined; // Optional Parameter
        this.age = undefined; // Optional Parameter
        this.address = undefined; // Optional Parameter
    }

    setRollNumber(rollNumber) {
        this.rollNumber = rollNumber;
        return this;
    }
    setAge(age) {
        this.age = age;
        return this;
    }
    setAddress(address) {
        this.address = address;
        return this;
    }

    build() {
        return new Student(this);
    }
}

class Student {
    constructor(builder) {
        this.name = builder.name;
        this.rollNumber = builder.rollNumber;
        this.age = builder.age;
        this.address = builder.address;
    }

    static get StudentBuilder() {
        return StudentBuilder;
    }
}

class Builder_Design_Pattern {
    static main(args) {
        const student = new Student.StudentBuilder("pradeep")
            .setAddress("dada")
            .setAge(123)
            .setAddress("dads")
            .build();

        // Is modification of student possible..
    }
}

Builder_Design_Pattern.main([]);
